File: internal/store/sqliteimpl/lineage.go
package sqliteimpl

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/sachncs/promptsheon/internal/lineage"
)

type LineageRepository struct{ db *sql.DB }

func NewLineageRepository(db *sql.DB) *LineageRepository { return &LineageRepository{db: db} }

var _ lineage.Repository = (*LineageRepository)(nil)

func (r *LineageRepository) GetGraph(ctx context.Context, capabilityID string) (*lineage.Graph, error) {
	rows, err := r.db.QueryContext(ctx, `SELECT parent_capability_id, parent_version, child_capability_id, child_version, source, recommendation_id, created_at, created_by, notes FROM lineage_edges WHERE capability_id = ? ORDER BY created_at ASC`, capabilityID)
	if err != nil {
		return nil, fmt.Errorf("lineage: query: %w", err)
	}
	defer func() { _ = rows.Close() }()
	g := &lineage.Graph{CapabilityID: capabilityID}
	for rows.Next() {
		var e lineage.Edge
		var source string
		var recommendationID sql.NullString
		if err := rows.Scan(&e.Parent.CapabilityID, &e.Parent.Version, &e.Child.CapabilityID, &e.Child.Version, &source, &recommendationID, &e.CreatedAt, &e.CreatedBy, &e.Notes); err != nil {
			return nil, fmt.Errorf("lineage: scan: %w", err)
		}
		e.Source = lineage.Source(source)
		if recommendationID.Valid {
			e.RecommendationID = recommendationID.String
		}
		g.Edges = append(g.Edges, e)
		g.UpdatedAt = e.CreatedAt
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return g, nil
}

func (r *LineageRepository) PutGraph(ctx context.Context, g *lineage.Graph) error {
	if g == nil {
		return errors.New("lineage: nil graph")
	}
	if err := g.Validate(); err != nil {
		return err
	}
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback() }()
	if _, err = tx.ExecContext(ctx, `DELETE FROM lineage_edges WHERE capability_id = ?`, g.CapabilityID); err != nil {
		return err
	}
	for i := range g.Edges {
		e := &g.Edges[i]
		var recommendationID any
		if e.RecommendationID != "" {
			recommendationID = e.RecommendationID
		}
		if _, err = tx.ExecContext(ctx, `INSERT INTO lineage_edges (capability_id, parent_capability_id, parent_version, child_capability_id, child_version, source, recommendation_id, created_at, created_by, notes) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`, g.CapabilityID, e.Parent.CapabilityID, e.Parent.Version, e.Child.CapabilityID, e.Child.Version, e.Source, recommendationID, e.CreatedAt, e.CreatedBy, e.Notes); err != nil {
			return fmt.Errorf("lineage: insert edge: %w", err)
		}
	}
	return tx.Commit()
}

File: packages/server/src/repos/lineage.ts
import type Database from 'better-sqlite3';
import type { LineageEdge } from '@promptsheon/shared';
import { notFound } from '@promptsheon/shared';

export class LineageRepo {
  constructor(private db: Database.Database) {}

  findById(id: number): LineageEdge | null {
    return this.db.prepare('SELECT * FROM lineage_edges WHERE id = ?').get(id) as LineageEdge | null;
  }

  findByCapabilityId(capabilityId: string): LineageEdge[] {
    return this.db.prepare('SELECT * FROM lineage_edges WHERE capability_id = ? ORDER BY created_at DESC')
      .all(capabilityId) as LineageEdge[];
  }

  findByParent(capabilityId: string, version: number): LineageEdge[] {
    return this.db.prepare('SELECT * FROM lineage_edges WHERE parent_capability_id = ? AND parent_version = ?')
      .all(capabilityId, version) as LineageEdge[];
  }

  findByChild(capabilityId: string, version: number): LineageEdge[] {
    return this.db.prepare('SELECT * FROM lineage_edges WHERE child_capability_id = ? AND child_version = ?')
      .all(capabilityId, version) as LineageEdge[];
  }

  create(data: {
    capabilityId: string; parentCapabilityId: string; parentVersion: number;
    childCapabilityId: string; childVersion: number; source: string;
    recommendationId?: string | null; createdBy?: string; notes?: string;
  }): LineageEdge {
    const now = new Date().toISOString();
    const result = this.db.prepare(`
      INSERT INTO lineage_edges (capability_id, parent_capability_id, parent_version, child_capability_id, child_version, source, recommendation_id, created_at, created_by, notes)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(data.capabilityId, data.parentCapabilityId, data.parentVersion, data.childCapabilityId, data.childVersion, data.source, data.recommendationId ?? null, now, data.createdBy ?? '', data.notes ?? '{}');
    return {
      id: Number(result.lastInsertRowid), capabilityId: data.capabilityId,
      parentCapabilityId: data.parentCapabilityId, parentVersion: data.parentVersion,
      childCapabilityId: data.childCapabilityId, childVersion: data.childVersion,
      source: data.source as LineageEdge['source'], recommendationId: data.recommendationId ?? null,
      createdAt: now, createdBy: data.createdBy ?? '', notes: data.notes ?? '{}',
    };
  }

  delete(id: number): void {
    this.db.prepare('DELETE FROM lineage_edges WHERE id = ?').run(id);
  }
}
